### PR TITLE
Add tests for poetry

### LINKED ISSUE #8: Introduce tests
We should introduce tests ensuring that each build behaves as it should:

- [ ] Has the `poetry` binary available
- [ ] The `poetry` version installed is the one in the tag
- [ ] The `poetry` installation works without virtual environments

### PR DESCRIPTION
**Pull request by gmargaritis**
Resolves #8

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -10,7 +10,23 @@ env:
   PLATFORMS: linux/amd64,linux/arm64/v8
 
 jobs:
+  test:
+    runs-on: ubuntu-20.04
```
Comment (parisk) [inline .github/workflows/release.yml:14]: Let's bump this to use `ubuntu-latest`

File: Dockerfile [hunk 2]
```diff
@@ -16,4 +16,6 @@ RUN python3 -m venv ${POETRY_HOME} &&\
     ${POETRY_HOME}/bin/pip install poetry==${POETRY_VERSION} &&\
     poetry config virtualenvs.create false
 
+COPY ./tests/tests.py /usr/src/app/tests.py
```
Comment (parisk) [inline Dockerfile:19]: We do not want to make this part of our Dockerfile. We should either create a separate build target, or bind mount it, when we `docker run`. I suggest the former.

**Review by parisk — CHANGES_REQUESTED** [review]
I would suggest taking a simpler approach on the CI level here, although the tests are  💯. A few suggestions:

- Run tests on the whole matrix on every push, so that we know our commits work as expected.
- If tests succeed, push image to GHCR tagged with `${github.sha}-{EXPECTED_RELEASE_TAG}` (e.g. `95db419-1.7.1-python-3.12-bookworm`)
- On release, tag the existing Git SHA tagged image on GHCR with the release version (e.g. `1.7.1-python-3.12-bookworm`)

**Comment by gmargaritis** [comment]
@parisk Let me know how to proceed from here. What changes do we want to make in the release workflow?

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+          - "3.8"
+          - "3.9"
+          - "3.10"
+          - "3.11"
+          - "3.12"
+        variant:
+          - bookworm
+          - slim
+    steps:
+      - uses: actions/checkout@v3
```
Comment (parisk) [inline .github/workflows/ci.yml:28]: Let's use the latest release: https://github.com/actions/checkout/releases/tag/v4.1.1

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+      - name: Build Docker Image
+        run: |
+          docker build --target test -t test-image \
+          --build-arg POETRY_VERSION=${{ env.POETRY_VERSION }} \
```
Comment (parisk) [inline .github/workflows/ci.yml:32]: We can just use the latest build push action with `push: false` and `target: test`.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+          .
+      - name: Run Tests in Docker
+        run: |
+          docker run --rm \
+          -e POETRY_VERSION=${{ env.POETRY_VERSION }} \
+          test-image python3 /usr/src/app/tests.py
+
+      - name: Log in to GitHub Container Registry
+        uses: docker/login-action@v1
```
Comment (parisk) [inline .github/workflows/ci.yml:41]: There's a more recent version

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        with:
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+      - name: Build and Push Docker Image
+        uses: docker/build-push-action@v3.1.0
```
Comment (parisk) [inline .github/workflows/ci.yml:47]: There is a more recent version: https://github.com/docker/build-push-action/releases/tag/v5.1.0

**Review by parisk — CHANGES_REQUESTED** [review]
I have added a few comments to tidy this up. We should also update the release workflow to just tag the images that already exist in the registry.

**Comment by gmargaritis** [comment]
@parisk How does the pulling of existing images based on the git sha work? Wouldn't the git sha be different for the commit that triggered the `ci` workflow, from the one that triggered the `release` workflow?

**Comment by parisk** [comment]
Nope, it's pretty much the same commit. All releases should happen on the `main` branch, for which the CI pipeline should always run: https://github.com/withlogicco/docker-poetry/pull/48/files#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03fR4-R6.

File: .github/workflows/release.yml [hunk 7]
```diff
@@ -29,58 +28,40 @@ jobs:
...
-          build-args: |
-            POETRY_VERSION=${{ env.POETRY_VERSION }}
-            PYTHON_VERSION=${{ matrix.python }}
-            VARIANT=${{ matrix.variant }}
+        run: |
+          ORIGINAL_TAG="${{ env.GHCR_IMAGE_REPOSITORY }}:${{ github.sha }}-${{ matrix.python }}-${{ matrix.variant }}"
+          NEW_TAG="${{ env.GHCR_IMAGE_REPOSITORY }}:${{ env.POETRY_VERSION }}-python-${{ matrix.python }}-${{ matrix.variant }}"
+          docker pull $ORIGINAL_TAG
+          docker tag $ORIGINAL_TAG $NEW_TAG
+          docker push $NEW_TAG
```
Comment (parisk) [inline .github/workflows/release.yml:43]: I suggest we use semantic variable names and reuse what we can

```suggestion
          TAG_SUFFIX="${{ matrix.python }}-${{ matrix.variant }}"
          GIT_SHA_TAG="${{ env.GHCR_IMAGE_REPOSITORY }}:${{ github.sha }}-${TAG_SUFFIX}"
          RELEASE_TAG="${{ env.GHCR_IMAGE_REPOSITORY }}:${{ env.POETRY_VERSION }}-python-${TAG_SUFFIX}"
          docker pull $GIT_SHA_TAG
          docker tag $GIT_SHA_TAG $RELEASE_TAG
          docker push $RELEASE_TAG
```

File: .github/workflows/release.yml [hunk 8]
```diff
@@ -29,58 +28,40 @@ jobs:
...
-          build-args: |
-            POETRY_VERSION=${{ env.POETRY_VERSION }}
-            PYTHON_VERSION=${{ matrix.python }}
-            VARIANT=${{ matrix.variant }}
+        run: |
+          ORIGINAL_TAG="${{ env.GHCR_IMAGE_REPOSITORY }}:${{ github.sha }}-${{ matrix.python }}-${{ matrix.variant }}"
+          NEW_TAG="${{ env.GHCR_IMAGE_REPOSITORY }}:${{ env.POETRY_VERSION }}-python-${{ matrix.python }}"
+          docker pull $ORIGINAL_TAG
+          docker tag $ORIGINAL_TAG $NEW_TAG
+          docker push $NEW_TAG
```
Comment (parisk) [inline .github/workflows/release.yml:51]: ditto: https://github.com/withlogicco/docker-poetry/pull/48/files#diff-87db21a973eed4fef5f32b267aa60fcee5cbdf03c67fafdc2a9b553bb0b15f34R39-R43

File: .github/workflows/release.yml [hunk 9]
```diff
@@ -29,58 +28,40 @@ jobs:
...
-          build-args: |
-            POETRY_VERSION=${{ env.POETRY_VERSION }}
-            PYTHON_VERSION=${{ matrix.python }}
-            VARIANT=${{ matrix.variant }}
+        run: |
+          ORIGINAL_TAG="${{ env.GHCR_IMAGE_REPOSITORY }}:${{ github.sha }}-${{ matrix.python }}-${{ matrix.variant }}"
+          NEW_TAG="${{ env.GHCR_IMAGE_REPOSITORY }}:${{ env.POETRY_VERSION }}-${{ matrix.variant }}"
+          docker pull $ORIGINAL_TAG
+          docker tag $ORIGINAL_TAG $NEW_TAG
+          docker push $NEW_TAG
```
Comment (parisk) [inline .github/workflows/release.yml:59]: ditto: https://github.com/withlogicco/docker-poetry/pull/48/files#diff-87db21a973eed4fef5f32b267aa60fcee5cbdf03c67fafdc2a9b553bb0b15f34R39-R43

File: .github/workflows/release.yml [hunk 10]
```diff
@@ -29,58 +28,40 @@ jobs:
...
-          build-args: |
-            POETRY_VERSION=${{ env.POETRY_VERSION }}
-            PYTHON_VERSION=${{ matrix.python }}
-            VARIANT=${{ matrix.variant }}
+        run: |
+          ORIGINAL_TAG="${{ env.GHCR_IMAGE_REPOSITORY }}:${{ github.sha }}-${{ matrix.python }}-${{ matrix.variant }}"
+          NEW_TAG="${{ env.GHCR_IMAGE_REPOSITORY }}:${{ env.POETRY_VERSION }}"
+          docker pull $ORIGINAL_TAG
+          docker tag $ORIGINAL_TAG $NEW_TAG
+          docker push $NEW_TAG
```
Comment (parisk) [inline .github/workflows/release.yml:67]: ditto: https://github.com/withlogicco/docker-poetry/pull/48/files#diff-87db21a973eed4fef5f32b267aa60fcee5cbdf03c67fafdc2a9b553bb0b15f34R39-R43

**Review by parisk — APPROVED** [review]
LGTM 🦌

**Pull request merged**